daml/Marketplace/Distribution/Auction/Utils.daml

```diff
@@ -0,0 +1,77 @@
+
+-- | A map/accumulator function for processing a submitted bid.
+-- Accepts the remaining quantity and price as an accululator along the current Bid as input
+-- Output is the updated accululator (if applicable) along with the Bid mapped to an Allocation (depending on the accumulator and bid information)
+dutchAuction : (Decimal, Decimal) -> Bid.T -> ((Decimal, Decimal), Auction.Allocation)
+dutchAuction acc@(0.0, _)   bid@Bid{..} = (acc, Auction.Allocation with quantity = 0.0; ..)
+dutchAuction (remaining, _) bid@Bid{..} =
+  let
+    quantity = details.quantity
+    bidPrice = details.price
+  in
+    if remaining >= quantity
+      then ((remaining - quantity, bidPrice), Auction.Allocation with ..)
+      else ((0.0, bidPrice), Auction.Allocation with quantity = remaining; ..)
+
+-- | An accumulator function which processes generates the settlement instruction for both an issuer and a bidder based of the output of the dutch auction
+generateSettlementInstruction : (Optional (ContractId AssetDeposit), [ContractId Settlement.SettlementInstruction]) -> Auction.SettleAllocation -> Update (Optional (ContractId AssetDeposit), [ContractId Settlement.SettlementInstruction])
+generateSettlementInstruction (None, _) _ = error "Auction:generateSettlementInstruction - Issuers AssetDeposit has no remaining units to allocate"
+generateSettlementInstruction (Some issuerDepositCid, settlementInstructionCids) Auction.SettleAllocation{..} = do
+    let
+      allocatedQuantity = allocation.quantity
+      bid = allocation.bid
+      allocatedAmount = allocatedQuantity * price
+      newStatus = if bid.details.quantity == allocatedQuantity
+                    then Bid.FullAllocation with ..
+                    else Bid.PartialAllocation with quantity = allocatedQuantity; ..
+
+    -- | Process bid side
+    (_, bidDepositCid) <- exerciseByKey @Bid.S (bid.assetId, bid.customer) Bid.ProcessAllocation with quantity = allocatedQuantity; amount = allocatedAmount; ..
+    let
+      buyer = bid.customer
+      payment = Settlement.SettlementDetails with senderAccount = bid.allocationAccount; depositCid = bidDepositCid; receiverAccount = issuerReceivableAccount
+
+    -- | Process issuer side
+    -- 1/ Extract their deposit
+    -- 2/ Split allocation quantity from this deposit, get the deposit matching the allocation. Can only result in either :
+        -- two asset deposits - One with the allocation, one with the remaining
+        -- one asset deposit - Occurs when processing the last bid when the auction is fully allocated
+    -- 3/ Create the settlement details for delivery of the asset to the buyer
+    issuerDeposit <- fetch issuerDepositCid
+    issuerDepositCids <- exercise issuerDepositCid AssetDeposit_Split with quantities = [allocatedQuantity]
+    let
+      seller = issuer
+      (allocatedDepositCid, remainingDepositCidOpt) = case issuerDepositCids of
+        (x :: [])      -> (x, None)
+        (x :: y :: []) -> (x, Some y)
```

Comment:
I saw the linter here suggesting `[x]` and [x, y]` instead.

Reply:
I thought this version was clearer than what the linter suggested but I can change it